### PR TITLE
More RDoc for Dir.glob

### PR DESCRIPTION
**Pull request by BurdetteLamar**
New globbing doc for other files to link to (in future PRs).

### CODE REVIEW

**Review by peterzhu2118 — COMMENTED** [review]
I took a quick look at this file, it covers a lot of topics and goes into quite a bit of details about the methods that perform globbing.

However, it's difficult for me to review because I'm not sure how it will be used. It's difficult for me to understand how much detail we should be going into and whether or not we should be talking about the globbing methods because I don't know what existing information it replaces.

It would make it easier for me to review if you could also edit the parts that this document replaces.

File: doc/globbing.rdoc [hunk 1]
```diff
@@ -0,0 +1,356 @@
...
+  flags = File::FNM_CASEFOLD | File::FNM_DOTMATCH
+
+==== Flag File::FNM_CASEFOLD
+
+The FNM_CASEFOLD flag specifies that the matching is to be case-insensitive:
+
+  Dir.glob('c*').take(5)
+  # => ["CONTRIBUTING.md", "COPYING", "COPYING.ja", "ccan", "class.c"]
+  Dir.glob('c*', flags: File::FNM_CASEFOLD).take(5)
+  # => ["CONTRIBUTING.md", "COPYING", "COPYING.ja", "ccan", "class.c"]
```
Comment (peterzhu2118) [inline doc/globbing.rdoc:227]: `FNM_CASEFOLD` is only useful for `File.fnmatch`, and not for any other methods (including `Dir.glob`). See the [documentation](https://docs.ruby-lang.org/en/master/File/Constants.html).

File: doc/globbing.rdoc [hunk 2]
```diff
@@ -0,0 +1,356 @@
...
+  Dir.glob('*', flags: File::FNM_DOTMATCH).take(5)
+  # => [".", ".appveyor.yml", ".cirrus.yml", ".dir-locals.el", ".document"]
+
+==== Flag File::FNM_EXTGLOB
+
+Flag File::FNM_EXTGLOB flag enables the pattern extension
+<tt>'{_a_,_b_}'</tt>, which matches pattern _a_ and pattern _b_;
+  behaves like
+  a {regexp union}[https://docs.ruby-lang.org/en/master/Regexp.html#method-c-union]
+  (e.g., <tt>'(?:_a_|_b_)'</tt>):
```
Comment (peterzhu2118) [inline doc/globbing.rdoc:245]: I don't think italics work in `<tt>` blocks, so the underscores will be generated.

File: doc/globbing.rdoc [hunk 3]
```diff
@@ -0,0 +1,356 @@
...
+
+Metacharacters <tt>'*'</tt> and <tt>'?'</tt> do not match directory separators.
+
+==== Flag File::FNM_SHORTNAME
+
+Patterns may match short names if they exist; Windows only.
+
+==== Flag File::FNM_SYSCASE
+
+Case sensitivity (sensitive or insensitive) is the same as the system.
```
Comment (peterzhu2118) [inline doc/globbing.rdoc:267]: We should note that it's also only useful for `File.fnmatch` and not any other methods.

File: doc/globbing.rdoc (hunk 2 above)
Comment (BurdetteLamar) [reply doc/globbing.rdoc:245]: The entire block is misaligned (displayed as code).   When that's corrected (next push), the italics work.

But the link there is fouled by `<tt>'{_a_,_b_}'</tt>` (as we've seen elsewhere), so we get `{regexp union}[https://docs.ruby-lang.org/en/master/Regexp.html#method-c-union]` in the output.  If I experimentally remove that text, the link is rendered correctly.

**Comment by BurdetteLamar** [comment]
> I took a quick look at this file, it covers a lot of topics and goes into quite a bit of details about the methods that perform globbing.
> 
> However, it's difficult for me to review because I'm not sure how it will be used. It's difficult for me to understand how much detail we should be going into and whether or not we should be talking about the globbing methods because I don't know what existing information it replaces.
> 
> It would make it easier for me to review if you could also edit the parts that this document replaces.

I share your concern.  When I started out, I was not sure where this was heading, but wanted at least to understand the implementations.

Now I'm thinking thus:

- Document completely at Dir.glob.
- Document completely at Pathname.glob;  "See" at Pathname#glob.
- Document completely at File.fnmatch;  "See" at Pathname#fnmatch.

Works for you?

**Comment by peterzhu2118** [comment]
Yes I agree. While duplicating the doc across several location isn't the ideal solution, I think it will be the easiest to understand for the reader.

**Comment by BurdetteLamar** [comment]
> Yes I agree. While duplicating the doc across several location isn't the ideal solution, I think it will be the easiest to understand for the reader.

I'm on it.

**Review by peterzhu2118 — COMMENTED** [review]
One minor comment, otherwise looks good!

File: dir.rb [hunk 4]
```diff
@@ -224,13 +224,22 @@ def self.[](*args, base: nil, sort: true)
...
-  # note that these are not regexps.
+  # Argument +patterns+ is a string pattern or an array of string patterns;
+  # note that these are not regexps; see below.
+  #
+  # Notes for the following examples:
+  #
+  # - <tt>'*'</tt> is the pattern that matches any entry name
+  #   except those that begin with <tt>'.'</tt>.
+  # - If optional keyword argument +base+ is not given,
+  #   the base directory whose entries are to be matched is <tt>'.'</tt>.
```
[the remark this answers is not in the document]
Comment (BurdetteLamar) [reply dir.rb]: Removed (as redundant);  minor edit to the other passage discussing +base+.

File: dir.rb [hunk 5]
```diff
@@ -248,9 +255,30 @@ def self.[](*args, base: nil, sort: true)
...
+  # Passing value +false+ with that keyword disables sorting
+  # (though the underlying file system may already have sorted the array).
+  #
+  # <b>Patterns</b>
+  #
   # Each pattern string is expanded
   # according to certain metacharacters;
-  # examples below use the
+  # examples below use the {Ruby file tree}[rdoc-ref:Dir@About+the+Examples]):
   # {Ruby project file tree}[https://github.com/ruby/ruby]:
```
Comment (peterzhu2118) [inline dir.rb:282]: There's sort of issue here, the `Ruby file tree` part is duplicated twice.

Comment (BurdetteLamar) [reply dir.rb]: Thanks for noticing; now fixed.

File: dir.rb [hunk 6]
```diff
@@ -248,10 +255,30 @@ def self.[](*args, base: nil, sort: true)
...
+  # Passing value +false+ with that keyword disables sorting
+  # (though the underlying file system may already have sorted the array).
+  #
+  # <b>Patterns</b>
+  #
   # Each pattern string is expanded
   # according to certain metacharacters;
-  # examples below use the
-  # {Ruby project file tree}[https://github.com/ruby/ruby]:
+  # examples below use the {Ruby file tree}[rdoc-ref:Dir@About+the+Examples]):
```
Comment (peterzhu2118) [inline dir.rb:281]: There's a trailing closing bracket on this line

```suggestion
  # examples below use the {Ruby file tree}[rdoc-ref:Dir@About+the+Examples]:
```

**Review by peterzhu2118 — APPROVED** [review]

**Pull request merged**